flink-clients/src/test/java/org/apache/flink/client/testjar/ClasspathProvider.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.client.testjar;
+
+import org.apache.flink.client.deployment.application.JarManifestParser;
+import org.apache.flink.client.program.PackagedProgram;
+
+import org.junit.rules.ExternalResource;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/** {@code ClasspathProvider} offers methods to generate classpaths based on actual jars. */
+public class ClasspathProvider extends ExternalResource {
+
+    private static final String CLASSPATH_PROPERTY_NAME = "java.class.path";
+
+    private static final Path TEST_JOB_JAR_PATH = Paths.get("target", "maven-test-jar.jar");
+
+    private static final Path JOB_JAR_PATH =
+            Paths.get("target", "maven-test-user-classloader-job-jar.jar");
+    private static final Path JOB_LIB_JAR_PATH =
+            Paths.get("target", "maven-test-user-classloader-job-lib-jar.jar");
+
+    private final TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    private File userDirHasTwoEntryClasses;
+
+    private File userDirHasEntryClass;
+
+    private File userDirHasNotEntryClass;
+
+    // this was added to support the legacy tests of ClassPathPackagedProgramRetrieverTest in the
+    // refactored PackagedProgramRetrieverImplTest
+    private File userDirWithTestJob;
+
+    private File userDirWithNoJarButTextFile;
+
+    private String originalSystemClasspath;
+
+    @Override
+    public void before() throws IOException {
+        temporaryFolder.create();
+
+        userDirHasTwoEntryClasses =
+                temporaryFolder.newFolder("_test_user_dir_has_two_entry_classes");
+        copyJar(JOB_JAR_PATH, userDirHasTwoEntryClasses);
```

Comment:
Why do we need these copy operations? Couldn't we construct the classpath from the original locations?

Reply:
`PackagedProgramRetrieverImpl.createEntryClassInformationProvider` requires a File instance referring to the folder containing the files. That's why I kept the logic for creating the folders as it already was in the original code base.